### PR TITLE
Efi 99

### PR DESCRIPTION
**Pull request by planckadmin**
In GitLab by @mb on Oct 17, 2023, 17:26

### CODE REVIEW

**Comment by planckdz** [comment]
Tests are not building.

**Comment by martinb89** [comment]
Tests are building now.

**Review by planckdz — CHANGES_REQUESTED** [review]

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/AuditLoggingUtil.swift [hunk 1]
```diff
@@ -11,8 +11,13 @@ import PlanckToolbox
 
 public protocol AuditLoggingUtilProtocol: AnyObject {
     
-    /// Save the log. If the file exceeds the time bound, the entries of the first day will be deleted.
-    func log(senderId: String, rating: String, maxLogTime: Int)
+    /// Save the log. 
+    /// If the file exceeds the time bound, we will keep only the entries of the last days according to the max log time passed.
+    func log(senderId: String, rating: String, maxLogTime: Int, errorCallback: @escaping (Error) -> Void)
+    
+    /// Save the event log (start, stop).
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/AuditLoggingUtil.swift:18]: It was needed to add start and stop events. As discussed iirc, it was when the app enter in foreground and background or terminates.

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/AuditLoggingUtil.swift [hunk 2]
```diff
@@ -11,8 +11,13 @@ import PlanckToolbox
...
+    /// If the file exceeds the time bound, we will keep only the entries of the last days according to the max log time passed.
+    func logEvent(maxLogTime: Int, auditLoggerEvent: AuditLoggerEvent, errorCallback: @escaping (Error) -> Void)
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/AuditLoggingUtil.swift:20]: ok, maxNumberOfDays first.

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/AuditLoggingUtil.swift [hunk 3]
```diff
@@ -11,8 +11,13 @@ import PlanckToolbox
...
 }
 
 public enum AuditLoggerEvent: String {
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/AuditLoggingUtil.swift:23]: It's not a general event.

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift [hunk 4]
```diff
@@ -82,118 +93,225 @@ public class FileExportUtil: NSObject, FileExportUtilProtocol {
...
+        auditLogQueue.addOperation { [weak self] in
+            guard let me = self else {
+                Log.shared.errorAndCrash("Lost myself")
                 return
             }
-            try data.write(to: fileUrl)
-            Log.shared.info("CSV file successfully saved")
-        } catch {
-            Log.shared.errorAndCrash(error: error)
+            var newCsv: String
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift:122]: ok

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift [hunk 5]
```diff
@@ -82,118 +93,225 @@ public class FileExportUtil: NSObject, FileExportUtilProtocol {
...
+            // If the file already exists, it has a signature. Let's verify it.
+            // If the signature is valid, the file will be persisted.
+            // Otherwise, something went wrong, and the user will be notified through the errorCallback.
+
+            // The persistedCsvContent is the content of the file that is already persisted, as it is, including the rows and the signature.
+            if let persistedCsvContent = me.getPersistedCSVContent() {
+                let verifyTextGroup = DispatchGroup()
+
+                // Craft the new CVS, that means add a row.
+                newCsv = me.createCSV(auditEventLog: auditEventLog,
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift:133]: It's used in :
me.prependSignatureAndSave(csv: newCsv, signature: signature, errorCallback: errorCallback)

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift [hunk 6]
```diff
@@ -82,118 +93,225 @@ public class FileExportUtil: NSObject, FileExportUtilProtocol {
...
+                                      maxLogTime: maxLogTime,
+                                      persistedCSVContent: persistedCsvContent)
+
+                // Verify the signature
+                let signatureToVerify = me.extractSignatureFrom(csv: persistedCsvContent)
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift:138]: It's needed the signature and the content without the signature separately.

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift [hunk 7]
```diff
@@ -82,118 +93,225 @@ public class FileExportUtil: NSObject, FileExportUtilProtocol {
...
+                let persistedCsvVersionWithoutSignature = me.removeSignatureFrom(csv: persistedCsvContent, signature: signatureToVerify)
+                verifyTextGroup.enter()
+                PEPSession().verifyText(persistedCsvVersionWithoutSignature, signature: signatureToVerify) { error in
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift]: ok

File: SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift [hunk 8]
```diff
@@ -9,51 +9,62 @@
...
+
     private let planckFolderName = "planck"
-    private let auditLogginggFileName = "auditLoggingg"
+    private let auditLoggingFileName = "auditLogging"
     private let csvExtension = "csv"
     private let commaSeparator = ","
     private let newLine = "\n"
     private var auditLoggingFilePath: URL?
-    
+    private let beginPGP = "-----BEGIN PGP MESSAGE----"
```
[the remark this answers is not in the document]
Comment (martinb89) [reply SubProjects/MessageModel/MessageModel/MessageModel/Util/FileExportUtil.swift:33]: Removed

**Review by planckdz — APPROVED** [review]

**Pull request merged**